### ISSUE TITLE
Ionic with Angular 9

### ISSUE THREAD

**Issue by radarsu** [nstudio/xplat#180]
I've been tracking issues #178 and #179.

I've managed to generate **angular 9 app with NativeScript**, but when I generate app **with Ionic it installs 8.2**.

Are there any plans to upgrade Ionic as well?

**Comment by NathanWalker** [comment]
Hi @radarsu yes this is underway 👍 We wanted to get initial workspace support released and Had to confirm ionic v9 support. By end of week this will be updated as well.

**Comment by johannesschobel** [comment]
Dear @NathanWalker ,
ok, if i want to start working on a new nx workspace with ionic support it's better to wait till the end of the week? Or is xplat v9 ready to use with ionic v5 / angular 9?
All the best

**Comment by NathanWalker** [comment]
@johannesschobel we’re updating the ionic generator to their latest versions with angular 9 support so once we release that support you can use it.

**Comment by johannesschobel** [comment]
Dear @NathanWalker ,
sorry to bother you.. Just wanted to ask about the latest status of the xplat angular 9 / ionic 5 release?

all the best,
Johannes

**Comment by NathanWalker** [comment]
@johannesschobel no worries, sorry last week was tough for sure. Going to try and wrap this up be end of day tomorrow - there was one Ivy build related issue with Ionic but should be able to clear.

**Comment by NathanWalker** [comment]
@johannesschobel There's still Ivy build errors with Ionic at moment unfortunately. Seems to be related to this I think https://github.com/angular/angular/issues/35338 -- If you know of any other tickets related to Ionic with Angular 9 please post back and can investigate more.

**Comment by johannesschobel** [comment]
Dear @NathanWalker , thanks for the update on this one.
The new Ionic5 uses Angular9 **with** ivy enabled right? So what is the main difference between the "default" ionic5-angular9-ivy application and the xplat application?

**Comment by NathanWalker** [comment]
I'll compare the 2 and see - not much difference - just brings modules in from workspace in Nx.

**Comment by johannesschobel** [comment]
Dear @NathanWalker ,
i vaguely remember, that - by the time of releasing ionic 5, i made a completely new `nrwl/nx` project to see if ionic5 is compatible. Then i installed `nstudio/xplat` and created a new ionic app (which would then be ionic4). Then i updated all dependencies to ionic5 and ran the app again. This worked for me.

However, i do not know, if this breaks something internally. Shall i create a demo repo?
All the best

**Comment by phillipplum** [comment]
I did the same thing as @johannesschobel. Everything also works as intended.

**Comment by johannesschobel** [comment]
Dear @phillipplum ,
however, i do not know, if this "simple update strategy" breaks something internally, which i have not found yet. That may be the issue @NathanWalker has , i guess :thinking:

**Comment by phillipplum** [comment]
> Dear @phillipplum ,
> however, i do not know, if this "simple update strategy" breaks something internally, which i have not found yet. That may be the issue @NathanWalker has , i guess 🤔

I totally agree @johannesschobel. I thought it was helpful that it worked without problems for me so far.

**Comment by johannesschobel** [comment]
Dear @NathanWalker , @phillipplum , and @radarsu ,

i have put up a `nrwl-xplat-ionic5` example repository to show that the default ionic5 application is running properly.
You can find the repository here:

https://github.com/johannesschobel/nrwl-xplat-ionic5-example

Just clone the repo and run `ng serve ionic-mobile` on your command line.

------------------------

Steps to reproduce:
- `npx create-nx-workspace@latest NAME`
  - with `empty` style
  - with `angular console`
- `ng add @nrwl/angular --defaults` to add angular
- `ng g @nrwl/angular:application todos` to add the default angular application

These steps are taken from the default `nrwl tutorial` page ( https://nx.dev/angular/tutorial/01-create-application )

Note that i updated the dependencies continuously
- `ng add @nstudio/xplat` to add xplat
As i still rely on the `ng` commands (instead of `nx`) i used the following command to generate a new application
- `ng generate @nstudio/xplat:app`
  - ionic application
  - angular frontend
  - i did **not** want to use the xplat architecture

Then i added the `capacitor` dependencies from `/apps/NAME/package.json` to the top-level `package.json` file and installed the latter via `npm install`.
Then i updated all dependencies to use the latest versions (i.e., update to ionic 5).

Hope this helps!?

**Comment by phillipplum** [comment]
Perfect @johannesschobel! I can reproduce that and no problems for me either.

**Comment by NathanWalker** [comment]
@johannesschobel that's awesome!  Super helpful - Also really exciting to hear that with Angular 9.1, Ivy can now use referenced packages again (as that was frustrating with 9.0) - Ok with your example repo I'm going to update the schematics to match a similar setup to allow this to be generated easier.

It's been bit touch/go for me given global situation that has made it harder for my wife and I since we have a 18 month old (with no nanny) and my daily work has been very busy so juggling a lot. 

Not sure if you had time to try hand at submitting a PR based on your sample repo to adjust schematic handlers?  Schematics can take a bit to get up to speed with and been thinking about doing some training sessions on adding features to the stack here.
This is a big priority and going to shoot to have updates posted by eod Tuesday (while working through the changes Monday). 

Hope everyone is staying safe and healthy out there.

**Comment by johannesschobel** [comment]
Dear @NathanWalker ,
first of all - thank you very much for your update. I am certainly aware, that the overall situation may be challenging for all of us. With this in mind, i really (!) appreciate your feedback and the effort you put in this project / product!

I stumbled upon the `schematics` approach (not only in `xplat`, but also in `nrwl` and `nestjs`), but never got the time to look into it. I am very sorry. Given this, i tried to make a "showcase repo" for you to follow my changes to make `nrwl-xplat-angular9-ionic5` work together :)

Hope you and your family stay healthy! All the best and keep up your good work! Looking forward to hear from you

**Comment by curiouscod3** [comment]
Please check. xplat changes angular-devkit version from 0.901 to 0.8XXX
https://github.com/nstudio/xplat/issues/186

**Comment by curiouscod3** [comment]
Dear @johannesschobel .

I've used your repo and add another ionic-mobile2 app with ng add. **with no supporting xplat architecture.** 
**Unfortunately, I got the same error : ng serve ionic-mobile2**
Your ionic-mobile version works: ng serve ionic-mobile.
I don't know why..

```
Uncaught ReferenceError: __importDefault is not defined
    at Module../src/app/app.component.ts (app.component.ts:10)
    at __webpack_require__ (bootstrap:84)
    at Module../src/app/app.module.ts (app.component.ts:13)
    at __webpack_require__ (bootstrap:84)
    at Module../src/main.ts (environment.ts:16)
    at __webpack_require__ (bootstrap:84)
    at Object.0 (main.ts:13)
    at __webpack_require__ (bootstrap:84)
    at checkDeferredModules (bootstrap:45)
    at Array.webpackJsonpCallback [as push] (bootstrap:32)
```

**Comment by johannesschobel** [comment]
Dear @curiouscod3 ,
that is because of the xplat schematics. As stated and discussed in this issue, the schematics **have not been adapted** to work for ionic5 yet!
The repository i have provided is a manual "hack", that shows, how it works and which dependencies have to be adapted to properly work with ionic5 / angular9.

Once the schematics are properly updated, it should work out of the box!

All the best

**Comment by phillipplum** [comment]
Hello,
maybe @johannesschobel or @NathanWalker have an idea. I'm not sure if it depends on manually changing the dependencies. I now have a normal angular app and an ionic app. If I now create a lib it can be loaded in the angular project, not in the ionic project.

./app/angular/tsconfig.json
```
{
  "extends": "../../tsconfig.json",
  "compilerOptions": {
    "types": ["node", "jest"]
  }
}
```

./app/ionic-app/tsconfig.json
```
{
  "compileOnSave": false,
  "compilerOptions": {
    "outDir": "./dist/out-tsc",
    "sourceMap": true,
    "declaration": false,
    "module": "esnext",
    "moduleResolution": "node",
    "emitDecoratorMetadata": true,
    "experimentalDecorators": true,
    "importHelpers": true,
    "target": "es2015",
    "typeRoots": ["node_modules/@types"],
    "lib": ["es2018", "dom"],
    "baseUrl": "./"
  }
}
```

And the ./tsconfig.json
```
{
  "compileOnSave": false,
  "compilerOptions": {
    "rootDir": ".",
    "sourceMap": true,
    "declaration": false,
    "moduleResolution": "node",
    "emitDecoratorMetadata": true,
    "experimentalDecorators": true,
    "importHelpers": true,
    "target": "es2015",
    "module": "esnext",
    "typeRoots": ["node_modules/@types"],
    "lib": ["es2017", "dom"],
    "skipLibCheck": true,
    "skipDefaultLibCheck": true,
    "baseUrl": ".",
    "paths": {
      "@mytestworkspace/test": ["libs/test/src/index.ts"],
      "@mytestworkspace/shared": ["libs/shared/src/index.ts"]
    }
  },
  "exclude": ["node_modules", "tmp"]
}
```

I have not yet found any other issue here in the github related to this issue and therefore i think that it is due to the manual changes? If not, please excuse me, I will create a new issue.

Stay healthy.

**Comment by johannesschobel** [comment]
maybe (!) it is, because the one compiles against es2017, the other one against es2018? But not sure..
In general, this should work, i guess, as ionic builds upon angular?!

**Comment by NathanWalker** [comment]
I've made some significant progress on [this PR](https://github.com/nstudio/xplat/pull/183) today   - I noticed that your example repo @johannesschobel didn't use xplat architecture which is where some Ivy trouble comes into play since @angular/common/http being a deep import (not direct level) does not get processed by 9.1's auto ngcc since it's symlink/reference package. I'm experimenting with a few solutions. 
Off the bat it's looking like using `@angular/common` as the only direct dep in the app itself (where all others are just references to root) and then adding a postinstall hook to ensure ngcc processes it may work:
```
ngcc -t @angular/common/http -l debug --create-ivy-entry-points
```

**Comment by NathanWalker** [comment]
Ok turns out the solution commented here actually fixes everything:
https://github.com/angular/angular/issues/25813#issuecomment-584866010
Adding this to the app's tsconfig path's:
```
"@angular/*": ["../../node_modules/@angular/*"]
```
Helps ensure ngcc picks up compiled files properly and things start working normally again.
I'll wrap the PR up later in the day and can finally get this published.

**Comment by NathanWalker** [comment]
This is published with `9.0.2` now. It also simplifies the package setup for Ionic app's in the workspace greatly.
There is one temporary thing to keep in mind until [this Nx related issue with pull request](https://github.com/nrwl/nx/pull/2811) is merged and released:

### Temporary Issue

If you created your workspace using Nrwl Nx cli vs. angular cli you will end up with `workspace.json` however there's a file utility in Nx that returns the wrong workspace filename depending on which 2 cli options you created your workspace with.

### Solution

Just renamed `workspace.json` to `angular.json` and you will be able to run your Ionic app in the workspace just fine. Again once the referenced PR is released this will no longer be the case.

### Reminder on how to setup Ionic app in Nx + xplat

```
nx g app
// or if you setup your workspace with angular cli:
ng g app
```

* Choose Ionic and any of the other options are up to you.
* After creation just be sure to run: `npm run clean.ionic.{your-app-name}` as this will ensure all dependencies are installed properly.
* You can then serve up your app for local in-browser development: `nx serve ionic.{your-app-name}`
* You can also prepare it for iOS/Android with Capacitor, for example:

```
cd apps/ionic-{your-app-name}
npm run cap.add.ios
npm run cap.add.android
```

**Comment by johannesschobel** [comment]
Dear @NathanWalker ,
first of all - **thank you very much for your work** in migrating xplat to the latest version!
I will try it this week - however, i will not be able to use the `xplat supporting architecture`.

All the best and thanks a lot!
